.pre-commit-config.yaml

```diff
@@ -1,45 +1,71 @@
 # Install pre-commit hooks via:
 # pre-commit install
 repos:
-- repo: https://github.com/pre-commit/pre-commit-hooks
-  rev: v3.4.0
-  hooks:
-    - id: double-quote-string-fixer
-    - id: end-of-file-fixer
-    - id: fix-encoding-pragma
-    - id: mixed-line-ending
-    - id: trailing-whitespace
-    - id: check-json
+  - repo: https://github.com/pre-commit/pre-commit-hooks
+    rev: v4.0.1
+    hooks:
+      - id: end-of-file-fixer
+      - id: fix-encoding-pragma
+      - id: mixed-line-ending
+      - id: trailing-whitespace
+      - id: check-json
+      - id: check-yaml
 
-- repo: https://github.com/pre-commit/mirrors-mypy
-  rev: v0.812
-  hooks:
-  - id: mypy
-    exclude: &exclude_files >
-      (?x)^(
-          docs/.*|
-      )$
+  - repo: https://github.com/mgedmin/check-manifest
+    rev: "0.46"
+    hooks:
+    - id: check-manifest
+      args: [--no-build-isolation]
+      additional_dependencies: [setuptools>=46.4.0, "reentry"]
 
-# yapf = yet another python formatter
-- repo: https://github.com/pre-commit/mirrors-yapf
-  rev: v0.30.0
-  hooks:
-  - id: yapf
-    name: yapf
-    args: ["-i"]
+  - repo: https://github.com/pycqa/isort
+    rev: 5.8.0
+    hooks:
+    - id: isort
 
-- repo: local
-  hooks:
-  - id: pylint
-    language: system
-    types: [file, python]
-    name: pylint
-    description: "This hook runs the pylint static code analyzer"
-    exclude: *exclude_files
-    entry: pylint
+  - repo: https://github.com/psf/black
+    rev: 21.5b2
+    hooks:
+    - id: black
 
-  - id: version-number
-    name: Check version numbers
-    entry: python ./.github/check_version.py
-    language: system
-    files: '^(setup.json)|(aiida_restapi/__init__.py)'
+  - repo: https://github.com/pre-commit/mirrors-mypy
+    rev: v0.812
+    hooks:
+    - id: mypy
+      additional_dependencies:
+      - pydantic~=1.8.2
+      exclude: >
+        (?x)^(
+            docs/.*|
+            tests/.*|
+            conftest.py
+        )$
+
+  - repo: https://github.com/PyCQA/pylint
+    rev: v2.8.3
+    hooks:
+    - id: pylint
+      additional_dependencies:
```

Comment:
Ah, I wasn't aware of this possibility!

I get that it is in nice to have the pylint pre-commit check in an isolated environment.
But as it is now it, this change is also not ideal since we now need to maintain the package dependencies in two places.

I'm not yet convinced that this is really better - can you convince me?

Reply:
I just feel it is the lesser of two evils. It kind of defeats the purpose of pre-commit, if you need to have system installs, since you can't use it as intended with `pre-commit install`, you can't use the pre-commit/action in the workflows, and/or you can't fully use https://pre-commit.ci/.
So I feel the hassle of keeping consistency, where it will be infrequent that dependencies change and it should anyhow be obvious when pylint fails, is a secondary concern to the frequent use of these pre-commit features.

I've looked for a specific solution but there does not appear to be one (https://github.com/pre-commit/pre-commit/issues/730). Of course you could think of some "meta" pre-commit hook to check this consistency (comparing setup.json and pre-commit yaml).

Note TBH I always use flake8 in my own projects instead of pylint and so never have this issue.

Reply:
> you can't use it as intended with `pre-commit install`

In practice it's not really an issue - the only difference is that instead of doing `pip install pre-commit` you do `pip install aiida-restapi[pre-commit]`

I get the point for the github actions / generic CI stuff.

Let's keep it like this; we can always go back if we feel it's a hassle to maintain these duplicate dependency lists.

Reply:
> In practice it's not really an issue - the only difference is that instead of doing pip install pre-commit you do pip install aiida-restapi[pre-commit]

ah yeh but thats only if you use git from the same python environment as you've installed this. That's certainly not always the case, from example I mainly use Github Desktop and just have pre-commit only installed in my root environment